kotlin/src/main/java/com/squareup/moshi/KotlinJsonAdapter.kt

```diff
@@ -0,0 +1,141 @@
+@file:Suppress("UNCHECKED_CAST")
+
+package com.squareup.moshi
+
+import com.squareup.moshi.ClassJsonAdapter.includeField
+import com.squareup.moshi.ClassJsonAdapter.isPlatformType
+import java.lang.reflect.Modifier
+import java.lang.reflect.Type
+import kotlin.reflect.KFunction
+import kotlin.reflect.KParameter
+import kotlin.reflect.KProperty1
+import kotlin.reflect.full.declaredMemberProperties
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.jvm.isAccessible
+import kotlin.reflect.jvm.javaField
+
+/**
+ * A [JsonAdapter.Factory] for Kotlin data classes. If this is added to Moshi,
+ * data classes will be treated differently from other classes.
+ *
+ * Moshi will call its primary constructor to create the object. This allows for delegated properties
+ * to be initialized correctly. If the JSON does not contain a value for a parameter with a default value
+ * the default value is used. If no default value is present an [IllegalArgumentException] will be
+ * thrown during deserialization.
+ *
+ * Data classes to be (de-)serialized must conform to the following rules:
+ *
+ * - Any properties not found in the primary constructor must be transient, including delegated properties.
+ * - The primary constructor may not contain transient properties unless a default value is provided.
+ * - If the class is a platform type (kotlin.*), it may not contain private fields.
```

Comment:
> Any properties not found in the primary constructor must be transient, including delegated properties.
Thinking about this, I don't think it's a necessary restriction. Instead, these fields could be deserialized/serialized just like the `ClassJsonAdapter` does currently using `FieldBinding`s. I'll lift this restriction.

> The primary constructor may not contain transient properties unless a default value is provided.

If it contained a transient property `p` without a default value, Moshi couldn't call the constructor since it may not provide values for `p` due to the `Transient` annotation. Thus, in order for the constructor call to succeed even though the `valuesByParam` don't contain an entry `(p, ...)`, the constructor will need to contain a default value for `p`.

> If the class is a platform type (kotlin.*), it may not contain private fields.

Same restriction as in the `ClassJsonAdapter`.